nussl/datasets/hooks.py

```diff
@@ -417,6 +427,146 @@ def process_item(self, item):
+                 balance: bool = False,
+                 **kwargs):
+        self.threshold_db = threshold_db
+        self.segment_dur = segment_dur
+        self.hop_ratio = hop_ratio
+        self.sample_rate = sample_rate
+        self.salient_src = salient_src
+        self.folder = folder
+        self.verbose = verbose
+        self.balance = balance
+
+        self.mix_src = MixSourceFolder(
+            folder, mix_folder, source_folders, ext, make_mix, **kwargs
+        )
+        self.song_metadata = self._populate_metadata()
+        self.offset = 0.0
+
+        super(SalientExcerptMixSourceFolder, self).__init__(
+            self._get_mix,
+            len(self.song_metadata),
+            **kwargs
+        )
+
+    def _populate_metadata(self):
+        """Makes a metadata dict of salient start times for each source."""
+
+        if self.verbose:
+            logging.info(f'Preprocessing data at {self.folder}.')
+
+        items = self.mix_src.get_items(self.folder)
+        balanced_samples = []
+        metadata = []
+        for item in tqdm.tqdm(items, disable=not self.verbose):
+            mix, sources = self.mix_src.get_mix_and_sources(item)
+            target_src = sources[self.salient_src]
+            target_audio = target_src.audio_data
+            salient_starts = utils.find_salient_starts(target_audio,
+                                                       self.segment_dur,
+                                                       self.hop_ratio,
+                                                       mix.sample_rate,
+                                                       self.threshold_db)
+            # this is a fairly cheap operation, no need to embed it within a
+            # conditional
+            balanced_samples.append((len(salient_starts), item,
+                                     salient_starts, mix.sample_rate))
+            for start in salient_starts:
+                if (start + self.segment_dur * mix.sample_rate) <= len(mix):
+                    metadata.append({
+                        'mixsrc_item': item,
+                        'start': start / mix.sample_rate
+                    })
+        # if the balance flag is set then balance the metadata
+        metadata = self._balance_set(balanced_samples) if self.balance \
+            else metadata
+        return metadata
+
+    @staticmethod
+    def _balance_set(sample_counts):
+        """
+        Balance the dataset to contain at least the average number of audio
+        samples
+        Args:
+            sample_counts: counts of the salient samples per audio files
+        Returns:
+        """
+        # TODO: Verify the method of balancing, option 1: oversample until
+        #  everything matches the max or option 2: oversample some under sample
+        #  others
+        avg_length = np.mean([elem[0] for elem in sample_counts])
+        metadata = []
+        for count, song, starts, sample_rate in sample_counts:
```

Comment:
Nit: the AudioSignal might not be a song, so perhaps this is a confusing variable name. `item` would be better and more consistent with the rest of the object.